### PR TITLE
Merge multiple parameters in newBPFFilter

### PR DESCRIPTION
**Pull request by Y7n05h**
Signed-off-by: Y7n05h <Y7n05h@protonmail.com>

### Short description

As discussed in #11526.
It is necessary to combine the parameters in `newBPFFilter` into a table.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

File: pdns/bpf-filter.cc [hunk 1]
```diff
@@ -312,18 +312,17 @@ static FDWrapper loadProgram(const struct bpf_insn* filter, size_t filterSize)
   return fd;
 }
 
-
-BPFFilter::BPFFilter(const BPFFilter::MapConfiguration& v4, const BPFFilter::MapConfiguration& v6, const BPFFilter::MapConfiguration& qnames, BPFFilter::MapFormat format, bool external): d_mapFormat(format), d_external(external)
+BPFFilter::BPFFilter(std::unordered_map<std::string, MapConfiguration> configs, BPFFilter::MapFormat format, bool external)
```
Comment (cmouse) [inline pdns/bpf-filter.cc:315]: you removed initializers from here, was that intentional?

File: pdns/bpf-filter.cc [hunk 2]
```diff
@@ -312,18 +312,17 @@ static FDWrapper loadProgram(const struct bpf_insn* filter, size_t filterSize)
...
 {
+
```
Comment (cmouse) [inline pdns/bpf-filter.cc:317]: stray whitespace

File: pdns/bpf-filter.cc [hunk 3]
```diff
@@ -312,18 +312,17 @@ static FDWrapper loadProgram(const struct bpf_insn* filter, size_t filterSize)
...
   if (d_mapFormat != BPFFilter::MapFormat::Legacy && !d_external) {
     throw std::runtime_error("Unsupported eBPF map format, the current internal implemenation only supports the legacy format");
   }
-
   auto maps = d_maps.lock();
 
-  maps->d_v4 = BPFFilter::Map(v4, d_mapFormat);
-  maps->d_v6 = BPFFilter::Map(v6, d_mapFormat);
-  maps->d_qnames = BPFFilter::Map(qnames, d_mapFormat);
+  maps->d_v4 = BPFFilter::Map(configs["v4Params"], d_mapFormat);
```
Comment (cmouse) [inline pdns/bpf-filter.cc:323]: d_mapFormat is now uninitialized here.

File: pdns/dnsdist-lua-bindings.cc [hunk 4]
```diff
@@ -425,40 +425,53 @@ void setupLuaBindings(LuaContext& luaCtx, bool client)
...
   using bpfFilterMapParams = boost::variant<uint32_t, LuaAssociativeTable<boost::variant<uint32_t, std::string>>>;
-  luaCtx.writeFunction("newBPFFilter", [client](bpfFilterMapParams v4Params, bpfFilterMapParams v6Params, bpfFilterMapParams qnameParams, boost::optional<bool> external) {
-      if (client) {
-        return std::shared_ptr<BPFFilter>(nullptr);
-      }
-
-      BPFFilter::MapConfiguration v4Config, v6Config, qnameConfig;
+  using bpfopts_t = LuaAssociativeTable<boost::variant<bool, std::string, bpfFilterMapParams>>;
+  luaCtx.writeFunction("newBPFFilter", [client](bpfopts_t opts) {
+    if (client) {
```
Comment (cmouse) [inline pdns/dnsdist-lua-bindings.cc:430]: pointless whitespace change.

File: pdns/dnsdist-lua-bindings.cc [hunk 5]
```diff
@@ -425,40 +425,53 @@ void setupLuaBindings(LuaContext& luaCtx, bool client)
...
+    std::unordered_map<std::string, BPFFilter::MapConfiguration> mapsConfig;
 
-      auto convertParamsToConfig = [](bpfFilterMapParams& params, BPFFilter::MapType type, BPFFilter::MapConfiguration& config) {
-        config.d_type = type;
-        if (params.type() == typeid(uint32_t)) {
-          config.d_maxItems = boost::get<uint32_t>(params);
+    auto convertParamsToConfig = [&](const std::string name, BPFFilter::MapType type) {
+      if (!opts.count(name))
+        return;
+      auto& tmp = opts.at(name);
```
Comment (cmouse) [inline pdns/dnsdist-lua-bindings.cc:438]: const auto

File: pdns/dnsdist-lua-bindings.cc [hunk 6]
```diff
@@ -425,40 +425,53 @@ void setupLuaBindings(LuaContext& luaCtx, bool client)
...
+      if (tmp.type() != typeid(bpfFilterMapParams))
+        throw std::runtime_error("params is invalid");
+      auto& params = boost::get<bpfFilterMapParams>(tmp);
+      BPFFilter::MapConfiguration config;
+      config.d_type = type;
+      if (params.type() == typeid(uint32_t)) {
+        config.d_maxItems = boost::get<uint32_t>(params);
+      }
+      else if (params.type() == typeid(LuaAssociativeTable<boost::variant<uint32_t, std::string>>)) {
+        auto map = boost::get<LuaAssociativeTable<boost::variant<uint32_t, std::string>>>(params);
```
Comment (cmouse) [inline pdns/dnsdist-lua-bindings.cc:448]: const auto map&?

**Review by cmouse — CHANGES_REQUESTED** [review]

File: pdns/bpf-filter.cc (hunk 1 above)
Comment (Y7n05h) [reply pdns/bpf-filter.cc:315]: Sorry, this was an oversight.

File: pdns/dnsdist-lua-bindings.cc [hunk 7]
```diff
@@ -425,39 +425,52 @@ void setupLuaBindings(LuaContext& luaCtx, bool client)
...
 
-      BPFFilter::MapConfiguration v4Config, v6Config, qnameConfig;
-
-      auto convertParamsToConfig = [](bpfFilterMapParams& params, BPFFilter::MapType type, BPFFilter::MapConfiguration& config) {
+      const auto convertParamsToConfig = [&](const std::string name, BPFFilter::MapType type) {
+        if (!opts.count(name))
+          return;
+        const auto& tmp = opts.at(name);
+        if (tmp.type() != typeid(bpfFilterMapParams))
+          throw std::runtime_error("params is invalid");
```
Comment (rgacogne) [inline pdns/dnsdist-lua-bindings.cc:440]:
```suggestion
        if (tmp.type() != typeid(bpfFilterMapParams)) {
          throw std::runtime_error("params is invalid");
        }
```

File: pdns/dnsdist-lua-bindings.cc [hunk 8]
```diff
@@ -425,39 +425,52 @@ void setupLuaBindings(LuaContext& luaCtx, bool client)
...
         return std::shared_ptr<BPFFilter>(nullptr);
       }
+      std::unordered_map<std::string, BPFFilter::MapConfiguration> mapsConfig;
 
-      BPFFilter::MapConfiguration v4Config, v6Config, qnameConfig;
-
-      auto convertParamsToConfig = [](bpfFilterMapParams& params, BPFFilter::MapType type, BPFFilter::MapConfiguration& config) {
+      const auto convertParamsToConfig = [&](const std::string name, BPFFilter::MapType type) {
+        if (!opts.count(name))
+          return;
```
Comment (rgacogne) [inline pdns/dnsdist-lua-bindings.cc:437]:
```suggestion
        if (!opts.count(name)) {
          return;
        }
```

File: pdns/dnsdist-lua-bindings.cc [hunk 9]
```diff
@@ -425,39 +425,52 @@ void setupLuaBindings(LuaContext& luaCtx, bool client)
...
       BPFFilter::MapFormat format = BPFFilter::MapFormat::Legacy;
-      if (external && *external) {
-        format = BPFFilter::MapFormat::WithActions;
+      bool external = false;
+      if (opts.count("external")) {
+        const auto& tmp = opts.at("external");
+        if (tmp.type() != typeid(bool))
+          throw std::runtime_error("params is invalid");
+        if ((external = boost::get<bool>(tmp)))
+          format = BPFFilter::MapFormat::WithActions;
```
Comment (rgacogne) [inline pdns/dnsdist-lua-bindings.cc:470]:
```suggestion
        if (tmp.type() != typeid(bool)) {
          throw std::runtime_error("params is invalid");
        }
        if ((external = boost::get<bool>(tmp))) {
          format = BPFFilter::MapFormat::WithActions;
        }
```

File: pdns/dnsdistdist/docs/reference/ebpf.rst [hunk 10]
```diff
@@ -13,27 +13,27 @@ These are all the functions, objects and methods related to the :doc:`../advance
...
   Maps can be pinned to a filesystem path, which makes their content persistent across restarts and allows external programs to read their content and to add new entries. dnsdist will try to load maps that are pinned to a filesystem path on startups, inheriting any existing entries, and fall back to creating them if they do not exist yet. Note that the user dnsdist is running under must have the right privileges to read and write to the given file, and to go through all the directories in the path leading to that file. The pinned path must be on a filesystem of type ``BPF``, usually below ``/sys/fs/bpf/``.
 
-  :param int maxV4: Maximum number of IPv4 entries in this filter
-  :param int maxV6: Maximum number of IPv6 entries in this filter
-  :param int maxQNames: Maximum number of QName entries in this filter
-
-  :param table v4Params: A table of options for the IPv4 filter map, see below
-  :param table v6Params: A table of options for the IPv6 filter map, see below
-  :param table qnameParams: A table of options for the qnames filter map, see below
+  :param table options: A table with key: value pairs with webserver options.
```
Comment (rgacogne) [inline pdns/dnsdistdist/docs/reference/ebpf.rst:26]:
```suggestion
  :param table options: A table with key: value pairs with options.
```

File: pdns/dnsdistdist/docs/reference/ebpf.rst [hunk 11]
```diff
@@ -13,27 +13,27 @@ These are all the functions, objects and methods related to the :doc:`../advance
...
 
   Options:
-
-  * ``maxItems``: int - The maximum number of entries in a given map. Default is 0 which will not allow any entry at all.
-  * ``pinnedPath``: str - The filesystem path this map should be pinned to.
+  * ``v4Params``: int or table - Maximum number of entries in this filter or a table with ``maxItems`` or ``pinnedPath`` .
```
Comment (rgacogne) [inline pdns/dnsdistdist/docs/reference/ebpf.rst:29]: I wonder if it wouldn't be easier for the three "*Params" to always take a table and drop the `int` option that can be specified via `maxItems` anyway.

Comment (Y7n05h) [reply pdns/dnsdistdist/docs/reference/ebpf.rst:29]: > I wonder if it wouldn't be easier for the three "*Params" to always take a table and drop the `int` option that can be specified via `maxItems` anyway.

Yes, I think that's a good idea. I will do that.

**Review by rgacogne — COMMENTED** [review]
Code looks great, thanks! A few nits on the documentation but nothing big.

**Comment by Y7n05h** [comment]
Thank you @cmouse and @rgacogne for your careful and rigorous Code Reviwe. my many oversights have added a lot of work for you. I will pay attention to improve the quality of code and documentation later.

File: pdns/dnsdistdist/docs/reference/ebpf.rst [hunk 12]
```diff
@@ -13,27 +13,28 @@ These are all the functions, objects and methods related to the :doc:`../advance
...
-.. function:: newBPFFilter(maxV4, maxV6, maxQNames) -> BPFFilter
-              newBPFFilter(v4Parameters, v6Parameters, qnamesParameters) -> BPFFilter
+.. function:: newBPFFilter(options) -> BPFFilter
+              newBPFFilter(v4Parameters, v6Parameters, qnamesParameters) -> BPFFilter (1.7.x)
+              newBPFFilter(maxV4, maxV6, maxQNames) -> BPFFilter (before 1.7.0)
 
   .. versionchanged:: 1.7.0
     This function now supports a table for each parameters, and the ability to use pinned eBPF maps.
+  .. versionchanged:: 1.8.0
+    This function now get the parameters via a table.
```
Comment (rgacogne) [inline pdns/dnsdistdist/docs/reference/ebpf.rst:23]:
```suggestion
    This function now gets its parameters via a table.
```

**Review by rgacogne — APPROVED** [review]
Great work, thanks!

**Pull request merged**